generator/src/test/java/com/scottlogic/deg/generator/fieldspecs/FieldSpecTests.java

```diff
@@ -0,0 +1,614 @@
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getDateTimeRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithGranularityRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        GranularityRestrictions restrictions = mock(GranularityRestrictions.class);
+        FieldSpec augmentedFieldSpec = original.withGranularityRestrictions(restrictions, FieldSpecSource.Empty);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getGranularityRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithFormatRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        FormatRestrictions restrictions = mock(FormatRestrictions.class);
+        FieldSpec augmentedFieldSpec = original.withFormatRestrictions(restrictions, FieldSpecSource.Empty);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getFormatRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithMustContainsRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        MustContainRestriction restrictions = mock(MustContainRestriction.class);
+        FieldSpec augmentedFieldSpec = original.withMustContainRestriction(restrictions);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getMustContainRestriction(), restrictions);
+    }
+
+    @Test
+    public void emptyFieldSpecsShouldBeEqualAndHaveSameHashCode(){
+        FieldSpec first = FieldSpec.Empty;
+        FieldSpec second = FieldSpec.Empty;
+
+        Assert.assertThat(first, equalTo(second));
+        Assert.assertThat(first.hashCode(), equalTo(second.hashCode()));
+    }
+
+    @Test
+    public void emptyFieldSpecAugmentedWithNothingShouldBeEqualToEmpty(){
+        FieldSpec first = FieldSpec.Empty;
+        FieldSpec augmented = FieldSpec.Empty.withNullRestrictions(null, FieldSpecSource.Empty);
+
+        Assert.assertThat(first, equalTo(augmented));
+        Assert.assertThat(first.hashCode(), equalTo(augmented.hashCode()));
+    }
+
+    @Test
+    public void fieldSpecsWithEqualSetRestrictionsShouldBeEqual(){
+        SetRestrictions aRestrictions = new MockSetRestrictions(true);
+        SetRestrictions bRestrictions = new MockSetRestrictions(true);
+        FieldSpec a = FieldSpec.Empty.withSetRestrictions(aRestrictions, FieldSpecSource.Empty);
+        FieldSpec b = FieldSpec.Empty.withSetRestrictions(bRestrictions, FieldSpecSource.Empty);
+
+        Assert.assertThat(a, equalTo(b));
+        Assert.assertThat(a.hashCode(), equalTo(b.hashCode()));
+    }
+
+    @Test
+    public void fieldSpecsWithUnequalSetRestrictionsShouldBeUnequal(){
+        SetRestrictions aRestrictions = new MockSetRestrictions(false);
```

Comment:
checking for not equal can be done with 

        SetRestrictions mockSetRestrictions1 =  mock(SetRestrictions.class);
        SetRestrictions mockSetRestrictions2 =  mock(SetRestrictions.class);

Reply:
these can even be set up at the start of the class, in a before if desired

Reply:
Unfortunately not, mockito doesn't support asserting/verifying `equals` or `hashCode`
https://github.com/mockito/mockito/wiki/FAQ

Reply:
This example is how you do it.

passing different mocks will not be equals, passing the same mock will be equals.

Reply:
this test is about having a bit more control of this, i think doing something like this allows for more control and clarity

Reply:
i believe comparing equality on things that are actually the same is much more clear than comparing stubbed overriden methods in 100 lines of private internal classes

 in this test, we don't need any further control than, is it the same or not,

Reply:
Complexity for complexity sake is bad, agreed.
However clarity in tests is important. 

We want to make sure we are calling `equals` and `hashCode` on the property types - like `SetRestrictions`. 

My view is the use of `mock(xxx1)` and `mock(xxx2)` would remove this clarity. With their use the tests would pass if `FieldSpec` used reference equality in its `equals` implementation where as it wouldn't with this implementation.